File: src/characters/butler.ml
open! Core
open! Import

include Character.Make (struct
    type t =
      { kind : Kind.t
      ; alignment : Alignment.t
      }

    let id = "butler"
    let name = "Butler"
    let t = { kind = Kind.Outsider; alignment = Alignment.Good }
    let kind { kind; _ } = kind
    let alignment { alignment; _ } = alignment
  end)

let night_action ~player_id:pid ~night:_ =
  Some
    (if_alive
       pid
       (let%bind.Botc_exec state = Botc_exec.get_state
        and () = Botc_exec.wake pid in
        let candidates = alive_except state pid in
        let%bind.Botc_exec _master = Botc_exec.ask pid "Who is your master?" candidates in
        Botc_exec.sleep pid))
;;

let day_action ~player_id:_ = None
let on_setup ~player_id:_ = None
let on_nominated ~player_id:_ ~nominator:_ = None
let on_executed ~player_id:_ = None
let on_night_kill ~player_id:_ = None

File: src/characters/butler.rs
//! Butler (Outsider): each night, choose a master. You may only vote when your
//! master votes. (Voting is a downstream concern; the Butler simply records the
//! master token.)

use crate::ability::Ability;
use crate::ctx::Ctx;
use crate::ids::PlayerId;
use crate::role::{Alignment, CharacterInfo, Kind};
use crate::token::Token;

character!(Butler, "butler", "Butler", Kind::Outsider, Alignment::Good);

impl Ability for Butler {
    fn info(&self) -> CharacterInfo {
        Self::INFO
    }

    fn on_night(&self, ctx: &mut Ctx, me: PlayerId, _night: u32) {
        if !ctx.grim.is_alive(me) {
            return;
        }
        ctx.wake(me);
        let targets = ctx.alive_others(me);
        let master = ctx.ask(me, "Choose your master", &targets);
        // Replace any previous master with tonight's choice.
        ctx.grim.remove_tokens_of_kind(me, Token::Master(me));
        ctx.grim.add_token(me, Token::Master(master));
        ctx.sleep(me);
    }
}
